design/src/rfcs/rfc0026_custom_validation_exceptions.md

```diff
@@ -0,0 +1,671 @@
+  that does not have `smithy.framework#ValidationException` attached will be
+  code generated so as to force the user to provide a [`core::ops::Fn`] value
+  that maps from the constrained operation's `ConstraintViolationExceptions`
+  container newtype to any modeled error shape attached to the operation.
+
+For example, an operation like:
+
+```smithy
+operation MyOperation {
+    input: MyOperationInput,
+    output: MyOperationOutput,
+    errors: [ValidationCustomError, OperationError1, ... OperationErrorN]
+}
+
+@error("client")
+structure ValidationCustomError {
+    @required
+    validationErrorMessage: String
+}
+```
+
+in a model generated with `codegenConfig.disableDefaultValidation` set to
+`true` will be implemented in user code as:
+
+```rust
+use my_server_sdk::{input, output, error}
+
+fn map_operation_validation_failure_to_error(errors: input::my_operation_input::ConstraintViolationExceptions) -> error::OperationError {
+    let mut validation_error_message: String::new();
+
+    for e in errors.into_iter() {
+        match e {
+            // ...
+        }
+    }
+
+    // We can return _any_ error attached to the operation.
+    error::OperationError::ValidationCustomError(
+        error::ValidationCustomError {
+            validation_error_message
+        }
+    )
+}
+
+async fn operation_handler(input: input::MyOperationInput) -> Result<output::MyOperationOutput, error::MyOperationError> {
+    // ...
+}
+```
+
+Since the operation does not have `smithy.framework#ValidationException`
+attached, when registering the handler in the service builder, the user must
+provide the mapping to an `error::MyOperationError` variant when constraint
+trait enforcement fails.
+
+```rust
+let my_service = MyService::builder_without_plugins()
+    .my_operation(operation_handler)
+    .on_my_operation_validation_failure(map_operation_validation_failure_to_error)
+    .build()
+    .unwrap();
+```
+
+This solution is ~~shamelessly pilfered~~ inspired by what the
+smithy-typescript server SDK implements, which is described [in this comment]
+and in [the documentation][smithy-typescript-custom-validation].
+
+Two last observations:
+
+1. The fact that this solution only allows service owners to map the constraint
+   violation exceptions to a modeled error shape, as opposed to giving them the
+   freedom of returning _anything_ that can be converted into an HTTP response
+   is _intentional_. Too much freedom would allow a user to return HTTP
+   responses that could break the underlying protocol.
+2. `ConstraintViolationException`s have until now been `pub(crate)`, for
+   exclusive use by the framework. Opting into `disableDefaultValidation`
+   should unhide only those associated to constrained operations where
+   `smithy.framework#ValidationException` is not attached to the operation. The
+   user shouldn't have a need to reach for the other operation's constraint
+   violation exceptions.
```

Comment:
Why do we care about this? It sounds like it wouldn't be a big deal that `ConstraintViolationException`s were public.

Reply:
As mentioned in the beginning of the document, this feature caters to a small tail of power users. I think making both `ConstraintViolation`s and `ConstraintViolationException`s public would unnecessarily expose API surface (i.e. make users overwhelmed when looking at Rust docs) when most users will just interact with `ConstraintViolation`s.